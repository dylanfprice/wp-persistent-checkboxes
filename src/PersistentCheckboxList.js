import md5 from 'md5'
import PropTypes from 'prop-types'
import React from 'react'

import {getValue, setValue} from './localStorage'

export default class PersistentCheckboxList extends React.Component {
    render () {
        const {labels} = this.props
        const listId = `list-${md5(labels.join('\n'))}`

        const checkboxes = labels.map(label => {
            const id = `${listId}-checkbox-${md5(label)}`
            return (
                <PersistentCheckbox
                    key={id}
                    id={id}
                    label={label}
                    persist={this.props.persist}
                />
            )
        })
        return (
            <div>
                {checkboxes}
            </div>
        )
    }
}

PersistentCheckboxList.propTypes = {
    labels: PropTypes.array.isRequired,
    persist: PropTypes.bool,
}

PersistentCheckboxList.defaultProps = {
    persist: true,
}

class PersistentCheckbox extends React.Component {
    constructor (props) {
        super(props)
        this.state = (
            this.props.persist
                ? {checked: getValue(this.getLocalStorageKey(), this.props.id)}
                : {checked: false}
        )
        this.onChange = this.onChange.bind(this)
    }

    getLocalStorageKey () {
        return 'checkbox-persist:' + window.location.pathname
    }

    onChange ({target: {checked}}) {
        if (this.props.persist) {
            setValue(this.getLocalStorageKey(), this.props.id, checked)
        }
        this.setState({checked: checked})
    }

    render () {
        const style = this.state.checked ? {textDecoration: 'line-through'} : {}
        return (
            <div>
                <input
                    id={this.props.id}
                    type="checkbox"
                    onChange={this.onChange}
                    checked={this.state.checked}
                />
                <label style={style} htmlFor={this.props.id}>
                    {this.props.label}
                </label>
            </div>
        )
    }
}

PersistentCheckbox.propTypes = {
    id: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    persist: PropTypes.bool,
}

PersistentCheckbox.defaultProps = {
    persist: true,
}
