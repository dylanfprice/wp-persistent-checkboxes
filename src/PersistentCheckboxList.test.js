import React from 'react';
import { shallow } from 'enzyme';

import PersistentCheckboxList from './PersistentCheckboxList';

describe( 'PersistentCheckboxList', () => {
	it( 'renders a list of checkboxes', () => {
		const labels = [ 'label one', 'label two', 'label three' ];
		const checkboxes = shallow(
			<PersistentCheckboxList labels={ labels } />
		);
		expect( checkboxes ).toMatchSnapshot();
	} );
} );
