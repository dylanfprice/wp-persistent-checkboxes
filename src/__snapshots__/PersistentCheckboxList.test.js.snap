// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersistentCheckboxList renders a list of checkboxes 1`] = `
<ol
  style={
    Object {
      "listStyleType": "none",
      "marginLeft": 0,
    }
  }
>
  <li
    key="list-331dfe0af8808b841db0c5ecf449572d-checkbox-65371417ffa624d6d9f3de3649ace43c"
  >
    <PersistentCheckbox
      id="list-331dfe0af8808b841db0c5ecf449572d-checkbox-65371417ffa624d6d9f3de3649ace43c"
      label={
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "label one",
            }
          }
        />
      }
    />
  </li>
  <li
    key="list-331dfe0af8808b841db0c5ecf449572d-checkbox-0e3c7175dad08d16072b98f692de3ae1"
  >
    <PersistentCheckbox
      id="list-331dfe0af8808b841db0c5ecf449572d-checkbox-0e3c7175dad08d16072b98f692de3ae1"
      label={
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "label two",
            }
          }
        />
      }
    />
  </li>
  <li
    key="list-331dfe0af8808b841db0c5ecf449572d-checkbox-eaafb0110b5c654b06ea4344a8a95b2b"
  >
    <PersistentCheckbox
      id="list-331dfe0af8808b841db0c5ecf449572d-checkbox-eaafb0110b5c654b06ea4344a8a95b2b"
      label={
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "label three",
            }
          }
        />
      }
    />
  </li>
</ol>
`;
